pkg/controller/installation/bpf.go

```diff
@@ -0,0 +1,170 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package installation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	operator "github.com/tigera/operator/api/v1"
+	crdv1 "github.com/tigera/operator/pkg/apis/crd.projectcalico.org/v1"
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/controller/migration/convert"
+	"github.com/tigera/operator/pkg/render"
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func bpfUpgradeDaemonsetEnvVar(r *ReconcileInstallation, ctx context.Context, install *operator.Installation, ds *appsv1.DaemonSet, fc *crdv1.FelixConfiguration, reqLogger logr.Logger) error {
+
+	// Query calico-node DS: if FELIX_BPFENABLED env var set and FC bpfEnabled unset then patch FC and quit.
+	patchFelixConfig, err := processDaemonsetEnvVar(r, ctx, ds, fc, reqLogger)
+	if err != nil {
+		return err
+	}
+
+	// Attempt to patch Felix Config now.
+	return patchFelixConfigurationImpl(r, ctx, install, fc, reqLogger, patchFelixConfig)
+}
+
+func bpfUpgradeWithoutDisruption(r *ReconcileInstallation, ctx context.Context, install *operator.Installation, ds *appsv1.DaemonSet, fc *crdv1.FelixConfiguration, reqLogger logr.Logger) error {
+
+	var patchFelixConfig bool
+
+	// Check the install dataplane mode is either Iptables or BPF.
+	installBpfEnabled := common.BpfDataplaneEnabled(&install.Spec)
+
+	// Check edge case where User has externally patched FelixConfig bpfEnabled which causes conflict to prevent Operator from upgrading dataplane.
+	if fc.Spec.BPFEnabled != nil {
+
+		fcBPFEnabled := *fc.Spec.BPFEnabled
+		if installBpfEnabled != fcBPFEnabled {
+
+			// Ensure Felix Config annotations are either empty or equal previous FC bpfEnabled value.
+			if fc.Annotations[render.BpfOperatorAnnotation] == strconv.FormatBool(installBpfEnabled) {
+				text := fmt.Sprintf("An error occurred while attempting patch Felix Config bpfEnabled: '%s' as Felix Config has been modified externally to '%s'",
+					strconv.FormatBool(installBpfEnabled),
+					strconv.FormatBool(fcBPFEnabled))
+				err := errors.New(text)
+				return err
+			}
+		}
+	}
+
+	if !installBpfEnabled {
+		// IP Tables dataplane:
+		// Only patch Felix Config once to prevent log spamming.
+		if fc.Spec.BPFEnabled == nil || *fc.Spec.BPFEnabled {
+			patchFelixConfig = true
+		}
+	} else {
+		// BPF dataplane:
+		// Check daemonset rollout complete before patching.
+		if fc.Spec.BPFEnabled == nil || !(*fc.Spec.BPFEnabled) {
+			patchFelixConfig = checkDaemonsetRolloutComplete(ds)
```

Comment:
I don't think we should be determining whether or not FelixConfiguration this far down the execution stack.

Can we formulate this so that the Reoncile() code is making the judgement call on which actions to take, and simple primitive helper functions perform common bits of work?

```go
if bpfEnabledOnDaemonSet(ds) && isAvailable(ds) {
  // The DaemonSet has BPF enabled, and all replicas are available. Ensure that 
  // FelixConfiguration has BPFEnalbed set to true.
  enableBPF()
}
```

^ This is pseudocode, but the pattern here is much more intuitive to me and I think results in more readable, maintainable code.

Reply:
Thanks @caseydavenport - understood - refactored accordingly - thanks